geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/TestGeoMesaDataStore.scala

```diff
@@ -0,0 +1,187 @@
+
+    override protected def rangeExact(row: Array[Byte]): TestRange = TestRange(row, IndexAdapter.rowFollowingRow(row))
+
+    override protected def scanPlan(sft: SimpleFeatureType,
+                                    ds: TestGeoMesaDataStore,
+                                    filter: TestFilterStrategyType,
+                                    hints: Hints,
+                                    ranges: Seq[TestRange],
+                                    ecql: Option[Filter]): TestQueryPlanType = TestQueryPlan(this, filter, ranges, ecql)
+  }
+
+  class TestAppendFeatureWriter(sft: SimpleFeatureType,
+                                ds: TestGeoMesaDataStore,
+                                indices: Option[Seq[TestFeatureIndexType]])
+      extends TestFeatureWriterType(sft, ds, indices) with TestAppendFeatureWriterType with TestFeatureWriter
+
+  class TestModifyFeatureWriter(sft: SimpleFeatureType,
+                                ds: TestGeoMesaDataStore,
+                                indices: Option[Seq[TestFeatureIndexType]],
+                                val filter: Filter)
+      extends TestFeatureWriterType(sft, ds, indices) with TestModifyFeatureWriterType with TestFeatureWriter
+
+  trait TestFeatureWriter extends TestFeatureWriterType {
+
+    override protected def createMutators(tables: IndexedSeq[String]): IndexedSeq[TestFeatureIndex] =
+      tables.map(t => ds.manager.indices(sft, IndexMode.Write).find(_.getTableName(sft.getTypeName, ds) == t).orNull)
+
+    override protected def executeWrite(mutator: TestFeatureIndex, writes: Seq[TestWrite]): Unit = {
+      writes.foreach { case TestWrite(row, feature, _) => mutator.features.add((row, feature)) }
+    }
+
+    override protected def executeRemove(mutator: TestFeatureIndex, removes: Seq[TestWrite]): Unit =
+      removes.foreach { case TestWrite(row, feature, _) => mutator.features.remove((row, feature)) }
+
+    override def wrapFeature(feature: SimpleFeature): TestWrappedFeature = TestWrappedFeature(feature)
+  }
+
+  case class TestQueryPlan(index: TestFeatureIndex,
+                           filter: TestFilterStrategyType,
+                           ranges: Seq[TestRange],
+                           ecql: Option[Filter]) extends TestQueryPlanType {
+    override def scan(ds: TestGeoMesaDataStore): CloseableIterator[SimpleFeature] = {
+      def contained(range: TestRange, row: Array[Byte]): Boolean =
+        byteComparator.compare(range.start, row) <= 0 && byteComparator.compare(range.end, row) > 0
+      index.features.toIterator.collect { case (row, sf) if ranges.exists(contained(_, row))  => sf }
+    }
+
+    override def explain(explainer: Explainer, prefix: String): Unit = {
+      explainer(s"ranges (${ranges.length}): ${ranges.take(5).map(r => s"[${r.start.mkString("")}:${r.end.mkString("")})")}")
+      explainer(s"ecql: ${ecql.map(org.locationtech.geomesa.filter.filterToString).getOrElse("INCLUDE")}")
```

Comment:
I started to feel like a horrible person for asking for this test given that you implemented an entire in-memory GeoMesa DataStore.  

The fact that you implemented an 'explain' method stopped that feeling completely.;)  Well done!